### PR TITLE
[SPARK-41457][PYTHON][TESTS] Refactor type annotations and dependency checks in tests

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR proposes to:

- Print out the correct error message when dependencies are not installed for `pyspark.sql.connect` in main code.
- Print out the correct message if 1. dependencies are missing or 2. build is not there in Spark Connect in tests.
- Ignore all MyPy errors from tests:
  - Initial intention was to annotate types for public APIs only.
  - Also removed unused ignores in other modules

Note that, for Spark Connect, we should follow both below to make sure they are properly skipped when dependences are not there, or jars do not exist.
- Always use `if should_test_connect` for `pyspark.sql.connect` and related dependency imports in tests.
- Use `@unittest.skipIf(not should_test_connect, connect_requirement_message)`

### Why are the changes needed?

- To make the development easier and faster.
- To make sure our Jenkins build pass without Python dependencies installed
  - See [Scaleaway](https://apache-spark.s3.fr-par.scw.cloud/index.html), see also https://github.com/apache/spark-website/blob/asf-site/developer-tools.md#test-coverage.

### Does this PR introduce _any_ user-facing change?

No, dev and test-only.

### How was this patch tested?

Manually tested without pandas, pyarrow and grpc, and verified the error messages.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @grundprinzip @hvanhovell @dongjoon-hyun @amaliujia @zhengruifeng @xinrong-meng FYI

**Comment by dongjoon-hyun** [comment]
Thank you, @HyukjinKwon !

**Review by dongjoon-hyun — COMMENTED** [review]
It seems that there is a leftover.
```
python/pyspark/sql/pandas/utils.py:83: error: unused "type: ignore" comment
[23] Found 1 error in 1 file (checked 373 source files)
```

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (if CI passes). Thank you so much for doing this, @HyukjinKwon !

**Review by zhengruifeng — APPROVED** [review]
LGTM, I am happy to see removing a lot of `  # type: ignore[import]`

**Review by grundprinzip — APPROVED** [review]
🚀 for fixing this!

**Comment by HyukjinKwon** [comment]
rebased

**Comment by dongjoon-hyun** [comment]
Could you check the linter failure?
```
starting mypy data test...
annotations failed data checks:
============================= test session starts ==============================
platform linux -- Python 3.9.5, pytest-7.1.3, pluggy-1.0.0
rootdir: /__w/spark/spark/python, configfile: ../dev/pyproject.toml
plugins: mypy-plugins-1.9.3
collected 40 items

python/pyspark/ml/tests/typing/test_classification.yml FF                [  5%]
python/pyspark/ml/tests/typing/test_clustering.yaml F                    [  7%]
python/pyspark/ml/tests/typing/test_evaluation.yml F                     [ 10%]
python/pyspark/ml/tests/typing/test_feature.yml FF                       [ 15%]
python/pyspark/ml/tests/typing/test_param.yml F                          [ 17%]
python/pyspark/ml/tests/typing/test_readable.yml F                       [ 20%]
python/pyspark/ml/tests/typing/test_regression.yml FFF                   [ 27%]
python/pyspark/sql/tests/typing/test_column.yml F                        [ 30%]
python/pyspark/sql/tests/typing/test_dataframe.yml FFFFFFF               [ 47%]
python/pyspark/sql/tests/typing/test_functions.yml F                     [ 50%]
python/pyspark/sql/tests/typing/test_readwriter.yml FF                   [ 55%]
python/pyspark/sql/tests/typing/test_session.yml FFFFF                   [ 67%]
python/pyspark/sql/tests/typing/test_udf.yml FFFFFFF                     [ 85%]
python/pyspark/tests/typing/test_context.yml F                           [ 87%]
python/pyspark/tests/typing/test_core.yml F                              [ 90%]
python/pyspark/tests/typing/test_rdd.yml FFF                             [ 97%]
python/pyspark/tests/typing/test_resultiterable.yml F                    [100%]
```

**Review by amaliujia — APPROVED** [review]

File: dev/lint-python [hunk 1]
```diff
@@ -104,7 +104,7 @@ function mypy_data_test {
       -c dev/pyproject.toml \
       --rootdir python \
       --mypy-only-local-stub \
-      --mypy-ini-file python/mypy.ini \
+      --mypy-ini-file dev/mypy.ini \
```
Comment (HyukjinKwon) [inline dev/lint-python:107]: Okay, MyPy requires the configuration file to locate in the root directory of the package. Let me move it back for now.

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
All python and linter tests passed. Merged to master.

**Comment by HyukjinKwon** [comment]
Thank you!

**Pull request closed without merging**